# 2º Limpieza de datos

In [844]:
# librerias
import pandas as pd
pd.set_option('display.max_columns', None)

import numpy as np

import warnings
warnings.filterwarnings('ignore')

## Limpieza de los datos de actores

In [845]:
actores=pd.read_csv('../data/actor.csv')

- Exploro los valores nulos y observamos que no hay

In [846]:
actores.isna().sum().sort_values(ascending=False) 



actor_id       0
first_name     0
last_name      0
last_update    0
dtype: int64

- Exploro los valores duplicados y observamos que tampoco hay

In [847]:
actores.duplicated().any() 

False

- Exploro los valores únicos

In [848]:
actores['actor_id'].unique() 
act_id = len(actores['actor_id'].unique())
act_id

200

In [849]:
actores['first_name'].unique() 
name1 = len(actores['first_name'].unique())
name1

128

In [850]:
actores['last_name'].unique() 
last1 = len(actores['last_name'].unique())
last1


121

Vemos que hay más nombres que apellidos, que puede tener sentido si alguna persona tiene el mismo apellido, por lo que decido combinar ambas tablas para simplificar 

In [851]:
actores["full_name"] = actores["first_name"] + " " + actores["last_name"]
actores["full_name"]

0         PENELOPE GUINESS
1            NICK WAHLBERG
2                 ED CHASE
3           JENNIFER DAVIS
4      JOHNNY LOLLOBRIGIDA
              ...         
195            BELA WALKEN
196             REESE WEST
197            MARY KEITEL
198          JULIA FAWCETT
199           THORA TEMPLE
Name: full_name, Length: 200, dtype: object

- Exploro los valores únicos - continuación

In [852]:
actores['last_update'].unique() # podemos observar que solo hay 1 valor único, es decir que los valores de cada fila son los mismos


array(['2006-02-15 04:34:33'], dtype=object)

- Decisión de borrado de columnas

In [853]:
actores.drop('last_update', axis=1, inplace=True) # borro la columna de last_update

In [854]:
actores = actores.drop(["first_name","last_name"], axis=1) # borro el nombre y apellido de cada actor ya que cree una columna conjunta

- Comprobación final

In [855]:
actores.head()

actor_id            full_name
0         1     PENELOPE GUINESS
1         2        NICK WAHLBERG
2         3             ED CHASE
3         4       JENNIFER DAVIS
4         5  JOHNNY LOLLOBRIGIDA

## Limpieza de los datos de categorías

In [856]:
categorias=pd.read_csv('../data/category.csv')

- Exploro los valores nulos y observamos que no hay

In [857]:
categorias.isna().sum().sort_values(ascending=False)


category_id    0
name           0
last_update    0
dtype: int64

- Exploro los valores duplicados y observamos que tampoco hay

In [858]:
categorias.duplicated().any() 

False

- Exploro los valores únicos

In [859]:
categorias['category_id'].unique() 
cat_id = len(categorias['category_id'].unique())
cat_id



16

In [860]:
categorias['name'].unique() 
cat_name = len(categorias['name'].unique())
cat_name





16

In [861]:
categorias['last_update'].unique() #igual que el caso anterior


array(['2006-02-15 04:46:27'], dtype=object)

- Decisión de borrado de columnas

In [862]:
categorias.drop('last_update', axis=1, inplace=True) # borro la columna de last_update

- Comprobación final

In [863]:
categorias.head()

category_id       name
0            1     Action
1            2  Animation
2            3   Children
3            4   Classics
4            5     Comedy

## Limpieza de los datos de películas

In [864]:
peliculas=pd.read_csv('../data/film.csv')


- Exploro los valores nulos y observamos que original_language_id tiene valores nulos

In [865]:
peliculas.isna().sum().sort_values(ascending=False)


original_language_id    1000
film_id                    0
title                      0
description                0
release_year               0
language_id                0
rental_duration            0
rental_rate                0
length                     0
replacement_cost           0
rating                     0
special_features           0
last_update                0
dtype: int64

- Exploro los valores duplicados y observamos que no hay

In [866]:
peliculas.duplicated().any() 

False

- Exploro algunos valores únicos

In [867]:
peliculas['film_id'].unique() 
film_id = len(peliculas['film_id'].unique())
film_id

1000

In [868]:
peliculas['title'].unique() 
title1  = len(peliculas['title'].unique())
title1 

1000

In [869]:
peliculas['language_id'].unique() 
language_id4  = len(peliculas['language_id'].unique())
language_id4

1

In [870]:
peliculas['release_year'].unique() 
release_year1 = len(peliculas['release_year'].unique())
release_year1

1

De aqui podemos sacar la siguiente info, en teoría tenemos mil títulos diferentes, los cuales solo están en 1 idioma, el inglés y además solo tienen el año de salida de 2006, lo cual parece mentira por lo que borraré también esa columna

- Decisión de borrado de columnas

In [871]:
peliculas.drop('last_update', axis=1, inplace=True) # borro la columna de last_update


In [872]:
peliculas.drop('original_language_id', axis=1, inplace=True) # borro la columna de original_language_id, ya que solo tenia valores nulos

In [873]:
peliculas.drop('release_year', axis=1, inplace=True) #esto parece mentira e inutil, lo quito tambn

- Comprobación final

In [874]:
peliculas.head()

film_id             title  \
0        1  ACADEMY DINOSAUR   
1        2    ACE GOLDFINGER   
2        3  ADAPTATION HOLES   
3        4  AFFAIR PREJUDICE   
4        5       AFRICAN EGG   

                                         description  language_id  \
0  A Epic Drama of a Feminist And a Mad Scientist...            1   
1  A Astounding Epistle of a Database Administrat...            1   
2  A Astounding Reflection of a Lumberjack And a ...            1   
3  A Fanciful Documentary of a Frisbee And a Lumb...            1   
4  A Fast-Paced Documentary of a Pastry Chef And ...            1   

   rental_duration  rental_rate  length  replacement_cost rating  \
0                6         0.99      86             20.99     PG   
1                3         4.99      48             12.99      G   
2                7         2.99      50             18.99  NC-17   
3                5         2.99     117             26.99      G   
4                6         2.99     130             22.99      G   

                   special_features  
0  Deleted Scenes,Behind the Scenes  
1           Trailers,Deleted Scenes  
2           Trailers,Deleted Scenes  
3    Commentaries,Behind the Scenes  
4                    Deleted Scenes

Tras verlo más limpio decido a futuro ponere el id de categoria y mover las categorias rental_duration	y rental_rate a otra tabla más adecuada, sea alquileres u otra

## Limpieza de los datos de inventarios

In [875]:
inventario=pd.read_csv('../data/inventory.csv')


- Exploro las columnas con valores nulos y observamos que no hay

In [876]:
inventario.isna().sum().sort_values(ascending=False)


inventory_id    0
film_id         0
store_id        0
last_update     0
dtype: int64

- Exploro los valores duplicados y observamos que tampoco hay

In [877]:
inventario.duplicated().any() 

False

- Exploro algunos valores únicos

In [878]:
inventario['inventory_id'].unique() 
inventory_id1 = len(inventario['inventory_id'].unique())
inventory_id1

1000

In [879]:
inventario['film_id'].unique() # tener en cuenta que en inventario no tenemos todas las peliculas
film_id2 = len(inventario['film_id'].unique())
film_id2

207

Interesante que el numero de film_id no coincide con el inventory_id, podría deberse a que tiene varias películas en inventario iguales??

- Decisión de borrado de columnas

In [880]:
inventario.drop('last_update', axis=1, inplace=True) # borro la columna de last_update

- Comprobación final

In [881]:
inventario.head()

inventory_id  film_id  store_id
0             1        1         1
1             2        1         1
2             3        1         1
3             4        1         1
4             5        1         2

## Limpieza de los datos de idiomas de las películas

In [882]:
idiomas=pd.read_csv('../data/language.csv')


- Exploro los valores nulos y observamos que no hay

In [883]:
idiomas.isna().sum().sort_values(ascending=False)

language_id    0
name           0
last_update    0
dtype: int64

- Exploro los valores duplicados y observamos que tampoco hay

In [884]:
idiomas.duplicated().any() 

False

- Exploro los valores únicos

In [885]:
idiomas['language_id'].unique() 
language_id1 = len(idiomas['language_id'].unique())
language_id1

6

In [886]:
idiomas['name'].unique() 
language_id1 = len(idiomas['name'].unique())
language_id1

6

- Decisión de borrado de columnas

In [887]:
idiomas.drop('last_update', axis=1, inplace=True) # borro la columna de last_update

- Comprobación final

In [888]:
idiomas.head()

language_id      name
0            1   English
1            2   Italian
2            3  Japanese
3            4  Mandarin
4            5    French

## Limpieza de los datos de disco duro

In [889]:
disco_duro_viejo=pd.read_csv('../data/old_HDD.csv')


- Exploro los valores nulos y observamos que no hay

In [890]:
disco_duro_viejo.isna().sum().sort_values(ascending=False)


first_name      0
last_name       0
title           0
release_year    0
category_id     0
dtype: int64

- Exploro los valores duplicados y observamos que tampoco hay

In [891]:
disco_duro_viejo.duplicated().any() 

False

- Cambio de columnas first_name y last_name por full_name igual que hice con actores

In [892]:
disco_duro_viejo["full_name"] = disco_duro_viejo["first_name"] + " " + disco_duro_viejo["last_name"]
disco_duro_viejo["full_name"]

0      PENELOPE GUINESS
1      PENELOPE GUINESS
2      PENELOPE GUINESS
3      PENELOPE GUINESS
4      PENELOPE GUINESS
             ...       
995        GOLDIE BRODY
996        GOLDIE BRODY
997        GOLDIE BRODY
998        GOLDIE BRODY
999        GOLDIE BRODY
Name: full_name, Length: 1000, dtype: object

- Exploro algunos valores únicos

In [893]:
disco_duro_viejo['release_year'].unique() #igual que el caso anterior, lo borraré
release_year2= len(disco_duro_viejo['release_year'].unique())
release_year2

1

In [894]:
disco_duro_viejo['category_id'].unique() 
category_id2 = len(disco_duro_viejo['category_id'].unique())
category_id2

16

In [895]:
disco_duro_viejo['title'].unique() #no tiene todas las peliculas
title2 = len(disco_duro_viejo['title'].unique())
title2

614

- Decisión de borrado de columnas

In [896]:
disco_duro_viejo.drop('release_year', axis=1, inplace=True) 

In [897]:
disco_duro_viejo = disco_duro_viejo.drop(["first_name","last_name"], axis=1) # borro el nombre y apellido de cada actor como antes

- Comprobación final

In [898]:
disco_duro_viejo.head()

title  category_id         full_name
0       ACADEMY DINOSAUR            6  PENELOPE GUINESS
1   ANACONDA CONFESSIONS            2  PENELOPE GUINESS
2            ANGELS LIFE           13  PENELOPE GUINESS
3  BULWORTH COMMANDMENTS           10  PENELOPE GUINESS
4          CHEAPER CLYDE           14  PENELOPE GUINESS

## Limpieza de los datos de alquileres

In [899]:
alquileres=pd.read_csv('../data/rental.csv')


- Exploro los valores nulos y observamos que no hay

In [900]:
alquileres.isna().sum().sort_values(ascending=False)

rental_id       0
rental_date     0
inventory_id    0
customer_id     0
return_date     0
staff_id        0
last_update     0
dtype: int64

- Exploro los valores duplicados y observamos que tampoco hay

In [901]:
alquileres.duplicated().any() 

False

- Exploro los valores únicos

In [902]:
alquileres['rental_id'].unique() 
rental_id1= len(alquileres['rental_id'].unique())
rental_id1

1000

In [903]:
alquileres['rental_date'].unique() 
rental_id1= len(alquileres['rental_date'].unique())
rental_id1

999

In [904]:
alquileres['inventory_id'].unique() 
inventory_id1= len(alquileres['inventory_id'].unique())
inventory_id1


1000

In [905]:
alquileres['customer_id'].unique() 
customer_id1= len(alquileres['customer_id'].unique())
customer_id1

485

In [906]:
alquileres['staff_id'].unique() 
staff_id1= len(alquileres['staff_id'].unique())
staff_id1

2

- Decisión de borrado de columnas

In [907]:
alquileres.drop('last_update', axis=1, inplace=True) # borro la columna de last_update

- Comprobación final

In [908]:
alquileres.head()

rental_id          rental_date  inventory_id  customer_id  \
0          1  2005-05-24 22:53:30           367          130   
1          2  2005-05-24 22:54:33          1525          459   
2          3  2005-05-24 23:03:39          1711          408   
3          4  2005-05-24 23:04:41          2452          333   
4          5  2005-05-24 23:05:21          2079          222   

           return_date  staff_id  
0  2005-05-26 22:04:30         1  
1  2005-05-28 19:40:33         1  
2  2005-06-01 22:12:39         1  
3  2005-06-03 01:43:41         2  
4  2005-06-02 04:33:21         1

# 3.1º Creación de base de datos 

### Formación de dataframes y cada columna de valores

Primero creo una tabla conjunta de películas e inventario, para tener los film_id e inventory_id juntos

In [909]:
peli_inv_temp = pd.merge(peliculas, inventario, on='film_id', how='inner') #utilizo un inner porque no tenemos todas las pelis y por tanto nos darian valores nulos
peli_inv_temp

film_id             title  \
0          1  ACADEMY DINOSAUR   
1          1  ACADEMY DINOSAUR   
2          1  ACADEMY DINOSAUR   
3          1  ACADEMY DINOSAUR   
4          1  ACADEMY DINOSAUR   
..       ...               ...   
995      222   DESERT POSEIDON   
996      222   DESERT POSEIDON   
997      222   DESERT POSEIDON   
998      223      DESIRE ALIEN   
999      223      DESIRE ALIEN   

                                           description  language_id  \
0    A Epic Drama of a Feminist And a Mad Scientist...            1   
1    A Epic Drama of a Feminist And a Mad Scientist...            1   
2    A Epic Drama of a Feminist And a Mad Scientist...            1   
3    A Epic Drama of a Feminist And a Mad Scientist...            1   
4    A Epic Drama of a Feminist And a Mad Scientist...            1   
..                                                 ...          ...   
995  A Brilliant Documentary of a Butler And a Fris...            1   
996  A Brilliant Documentary of a Butler And a Fris...            1   
997  A Brilliant Documentary of a Butler And a Fris...            1   
998  A Fast-Paced Tale of a Dog And a Forensic Psyc...            1   
999  A Fast-Paced Tale of a Dog And a Forensic Psyc...            1   

     rental_duration  rental_rate  length  replacement_cost rating  \
0                  6         0.99      86             20.99     PG   
1                  6         0.99      86             20.99     PG   
2                  6         0.99      86             20.99     PG   
3                  6         0.99      86             20.99     PG   
4                  6         0.99      86             20.99     PG   
..               ...          ...     ...               ...    ...   
995                4         4.99      64             27.99      R   
996                4         4.99      64             27.99      R   
997                4         4.99      64             27.99      R   
998                7         2.99      76             24.99  NC-17   
999                7         2.99      76             24.99  NC-17   

                     special_features  inventory_id  store_id  
0    Deleted Scenes,Behind the Scenes             1         1  
1    Deleted Scenes,Behind the Scenes             2         1  
2    Deleted Scenes,Behind the Scenes             3         1  
3    Deleted Scenes,Behind the Scenes             4         1  
4    Deleted Scenes,Behind the Scenes             5         2  
..                                ...           ...       ...  
995        Trailers,Behind the Scenes           996         2  
996        Trailers,Behind the Scenes           997         2  
997        Trailers,Behind the Scenes           998         2  
998                    Deleted Scenes           999         2  
999                    Deleted Scenes          1000         2  

[1000 rows x 12 columns]

In [910]:
peli_inv_temp.isna().sum().sort_values(ascending=False)

film_id             0
title               0
description         0
language_id         0
rental_duration     0
rental_rate         0
length              0
replacement_cost    0
rating              0
special_features    0
inventory_id        0
store_id            0
dtype: int64

Luego a partir de esa tabla, creo otras 2 tablas temporales una para alquileres y otra para inventario

In [911]:
rental_dur_rate_alquiler =  peli_inv_temp.drop(['film_id', 'title', 'description', 'language_id', 'length', 'replacement_cost', 'rating', 'special_features', 'store_id'], axis=1, inplace=False) # inplace false devuelve una tabla nueva en vez de modificar la original del drop
replacement_cost_inv = peli_inv_temp.drop(['title', 'description', 'language_id', 'rental_duration', 'rental_rate', 'length', 'rating', 'special_features', 'store_id', 'film_id'], axis=1, inplace=False) 


In [912]:
rental_dur_rate_alquiler.isna().sum().sort_values(ascending=False)

rental_duration    0
rental_rate        0
inventory_id       0
dtype: int64

In [913]:
replacement_cost_inv.isna().sum().sort_values(ascending=False)

replacement_cost    0
inventory_id        0
dtype: int64

In [914]:
new_invetory = pd.merge(inventario, replacement_cost_inv, on='inventory_id', how='left')
new_invetory 

inventory_id  film_id  store_id  replacement_cost
0               1        1         1             20.99
1               2        1         1             20.99
2               3        1         1             20.99
3               4        1         1             20.99
4               5        1         2             20.99
..            ...      ...       ...               ...
995           996      222         2             27.99
996           997      222         2             27.99
997           998      222         2             27.99
998           999      223         2             24.99
999          1000      223         2             24.99

[1000 rows x 4 columns]

In [915]:
new_invetory.isna().sum().sort_values(ascending=False)

inventory_id        0
film_id             0
store_id            0
replacement_cost    0
dtype: int64

Lo que quería hacer y he conseguido es meter el valor replacement_cost en inventario y rental_duration y rental_rate en alquileres. Esto podría darme errores a futuro al valorar datos, pero lo que se me ha ocurrido para obtener un resultado más o menos fiable será en MySQL obtener la media y/o mediana de estos valores (rental_duration y rental_rate)

In [916]:
new_alquiler=pd.merge(alquileres, rental_dur_rate_alquiler, on='inventory_id', how='inner') 
new_alquiler

rental_id          rental_date  inventory_id  customer_id  \
0            1  2005-05-24 22:53:30           367          130   
1           16  2005-05-25 00:43:11           389          316   
2           17  2005-05-25 01:06:36           830          575   
3           21  2005-05-25 01:59:46           146          388   
4           22  2005-05-25 02:19:23           727          509   
..         ...                  ...           ...          ...   
207        972  2005-05-30 20:21:07             2          411   
208        978  2005-05-30 21:30:52           407          154   
209        983  2005-05-30 22:15:51            63          115   
210        987  2005-05-30 22:59:12           390           11   
211        995  2005-05-31 00:06:02            17          150   

             return_date  staff_id  rental_duration  rental_rate  
0    2005-05-26 22:04:30         1                7         2.99  
1    2005-05-26 04:42:11         2                6         4.99  
2    2005-05-27 00:43:36         1                7         2.99  
3    2005-05-26 01:01:46         2                5         4.99  
4    2005-05-26 04:52:23         2                5         4.99  
..                   ...       ...              ...          ...  
207  2005-06-06 00:36:07         1                6         0.99  
208  2005-06-07 16:22:52         1                3         0.99  
209  2005-06-02 22:56:51         1                6         0.99  
210  2005-06-07 20:56:12         1                6         4.99  
211  2005-06-06 02:30:02         2                5         2.99  

[212 rows x 8 columns]

Hago inner para limpiar la tabla a solo los inventario disponibles, hay que tener en cuanta que antes al mirar valores únicos de inventario, los film_id eran 207 y ahora tenemos 212 filas, por lo que únicamente se repetirían 5 películas, el resto son únicas

In [917]:
new_alquiler.isna().sum().sort_values(ascending=False)

rental_id          0
rental_date        0
inventory_id       0
customer_id        0
return_date        0
staff_id           0
rental_duration    0
rental_rate        0
dtype: int64

Termino de mover las columnas que quería para tener los dataframes finales. En este caso quiero tener el category_id en peliculas directamente

In [918]:
disco_temp =  disco_duro_viejo.drop('full_name', axis=1, inplace=False) 


In [919]:
new_peliculas=pd.merge(peliculas, disco_temp, on='title', how='left')
new_peliculas

film_id              title  \
0           1   ACADEMY DINOSAUR   
1           1   ACADEMY DINOSAUR   
2           1   ACADEMY DINOSAUR   
3           1   ACADEMY DINOSAUR   
4           2     ACE GOLDFINGER   
...       ...                ...   
1381      997         YOUTH KICK   
1382      997         YOUTH KICK   
1383      998       ZHIVAGO CORE   
1384      999  ZOOLANDER FICTION   
1385     1000          ZORRO ARK   

                                            description  language_id  \
0     A Epic Drama of a Feminist And a Mad Scientist...            1   
1     A Epic Drama of a Feminist And a Mad Scientist...            1   
2     A Epic Drama of a Feminist And a Mad Scientist...            1   
3     A Epic Drama of a Feminist And a Mad Scientist...            1   
4     A Astounding Epistle of a Database Administrat...            1   
...                                                 ...          ...   
1381  A Touching Drama of a Teacher And a Cat who mu...            1   
1382  A Touching Drama of a Teacher And a Cat who mu...            1   
1383  A Fateful Yarn of a Composer And a Man who mus...            1   
1384  A Fateful Reflection of a Waitress And a Boat ...            1   
1385  A Intrepid Panorama of a Mad Scientist And a B...            1   

      rental_duration  rental_rate  length  replacement_cost rating  \
0                   6         0.99      86             20.99     PG   
1                   6         0.99      86             20.99     PG   
2                   6         0.99      86             20.99     PG   
3                   6         0.99      86             20.99     PG   
4                   3         4.99      48             12.99      G   
...               ...          ...     ...               ...    ...   
1381                4         0.99     179             14.99  NC-17   
1382                4         0.99     179             14.99  NC-17   
1383                6         0.99     105             10.99  NC-17   
1384                5         2.99     101             28.99      R   
1385                3         4.99      50             18.99  NC-17   

                             special_features  category_id  
0            Deleted Scenes,Behind the Scenes          6.0  
1            Deleted Scenes,Behind the Scenes          6.0  
2            Deleted Scenes,Behind the Scenes          6.0  
3            Deleted Scenes,Behind the Scenes          6.0  
4                     Trailers,Deleted Scenes         11.0  
...                                       ...          ...  
1381               Trailers,Behind the Scenes         12.0  
1382               Trailers,Behind the Scenes         12.0  
1383                           Deleted Scenes         11.0  
1384                  Trailers,Deleted Scenes          NaN  
1385  Trailers,Commentaries,Behind the Scenes          NaN  

[1386 rows x 11 columns]

In [920]:
new_peliculas['category_id'] = new_peliculas['category_id'].fillna(0)

In [921]:
new_peliculas['category_id'] = new_peliculas['category_id'].astype(int)

In [922]:
new_peliculas.isna().sum().sort_values(ascending=False)

film_id             0
title               0
description         0
language_id         0
rental_duration     0
rental_rate         0
length              0
replacement_cost    0
rating              0
special_features    0
category_id         0
dtype: int64

Actualizo los nombres originales para cada tabla

In [923]:
peliculas=new_peliculas
peliculas.drop(['replacement_cost','rental_duration', 'rental_rate'], axis=1, inplace=True)
peliculas

film_id              title  \
0           1   ACADEMY DINOSAUR   
1           1   ACADEMY DINOSAUR   
2           1   ACADEMY DINOSAUR   
3           1   ACADEMY DINOSAUR   
4           2     ACE GOLDFINGER   
...       ...                ...   
1381      997         YOUTH KICK   
1382      997         YOUTH KICK   
1383      998       ZHIVAGO CORE   
1384      999  ZOOLANDER FICTION   
1385     1000          ZORRO ARK   

                                            description  language_id  length  \
0     A Epic Drama of a Feminist And a Mad Scientist...            1      86   
1     A Epic Drama of a Feminist And a Mad Scientist...            1      86   
2     A Epic Drama of a Feminist And a Mad Scientist...            1      86   
3     A Epic Drama of a Feminist And a Mad Scientist...            1      86   
4     A Astounding Epistle of a Database Administrat...            1      48   
...                                                 ...          ...     ...   
1381  A Touching Drama of a Teacher And a Cat who mu...            1     179   
1382  A Touching Drama of a Teacher And a Cat who mu...            1     179   
1383  A Fateful Yarn of a Composer And a Man who mus...            1     105   
1384  A Fateful Reflection of a Waitress And a Boat ...            1     101   
1385  A Intrepid Panorama of a Mad Scientist And a B...            1      50   

     rating                         special_features  category_id  
0        PG         Deleted Scenes,Behind the Scenes            6  
1        PG         Deleted Scenes,Behind the Scenes            6  
2        PG         Deleted Scenes,Behind the Scenes            6  
3        PG         Deleted Scenes,Behind the Scenes            6  
4         G                  Trailers,Deleted Scenes           11  
...     ...                                      ...          ...  
1381  NC-17               Trailers,Behind the Scenes           12  
1382  NC-17               Trailers,Behind the Scenes           12  
1383  NC-17                           Deleted Scenes           11  
1384      R                  Trailers,Deleted Scenes            0  
1385  NC-17  Trailers,Commentaries,Behind the Scenes            0  

[1386 rows x 8 columns]

In [924]:
alquileres=new_alquiler
alquileres

rental_id          rental_date  inventory_id  customer_id  \
0            1  2005-05-24 22:53:30           367          130   
1           16  2005-05-25 00:43:11           389          316   
2           17  2005-05-25 01:06:36           830          575   
3           21  2005-05-25 01:59:46           146          388   
4           22  2005-05-25 02:19:23           727          509   
..         ...                  ...           ...          ...   
207        972  2005-05-30 20:21:07             2          411   
208        978  2005-05-30 21:30:52           407          154   
209        983  2005-05-30 22:15:51            63          115   
210        987  2005-05-30 22:59:12           390           11   
211        995  2005-05-31 00:06:02            17          150   

             return_date  staff_id  rental_duration  rental_rate  
0    2005-05-26 22:04:30         1                7         2.99  
1    2005-05-26 04:42:11         2                6         4.99  
2    2005-05-27 00:43:36         1                7         2.99  
3    2005-05-26 01:01:46         2                5         4.99  
4    2005-05-26 04:52:23         2                5         4.99  
..                   ...       ...              ...          ...  
207  2005-06-06 00:36:07         1                6         0.99  
208  2005-06-07 16:22:52         1                3         0.99  
209  2005-06-02 22:56:51         1                6         0.99  
210  2005-06-07 20:56:12         1                6         4.99  
211  2005-06-06 02:30:02         2                5         2.99  

[212 rows x 8 columns]

Por último voy a cambiarle el nombre del disco_duro_viejo a reparto, eliminar el category_id, meter los id de peliculas y actores y después borrar el full_name y el title, para dejar creado la tabla intermedia n-n de ambas

In [925]:
reparto=disco_duro_viejo
reparto.drop('category_id', axis=1, inplace=True)
reparto

title         full_name
0         ACADEMY DINOSAUR  PENELOPE GUINESS
1     ANACONDA CONFESSIONS  PENELOPE GUINESS
2              ANGELS LIFE  PENELOPE GUINESS
3    BULWORTH COMMANDMENTS  PENELOPE GUINESS
4            CHEAPER CLYDE  PENELOPE GUINESS
..                     ...               ...
995      COMANCHEROS ENEMY      GOLDIE BRODY
996        DAISY MENAGERIE      GOLDIE BRODY
997        DESERT POSEIDON      GOLDIE BRODY
998         EVERYONE CRAFT      GOLDIE BRODY
999         EXORCIST STING      GOLDIE BRODY

[1000 rows x 2 columns]

In [926]:
new_reparto1=pd.merge(actores, reparto, on='full_name', how='left')
new_reparto1

actor_id         full_name                  title
0            1  PENELOPE GUINESS       ACADEMY DINOSAUR
1            1  PENELOPE GUINESS   ANACONDA CONFESSIONS
2            1  PENELOPE GUINESS            ANGELS LIFE
3            1  PENELOPE GUINESS  BULWORTH COMMANDMENTS
4            1  PENELOPE GUINESS          CHEAPER CLYDE
...        ...               ...                    ...
1156       196       BELA WALKEN                    NaN
1157       197        REESE WEST                    NaN
1158       198       MARY KEITEL                    NaN
1159       199     JULIA FAWCETT                    NaN
1160       200      THORA TEMPLE                    NaN

[1161 rows x 3 columns]

In [927]:
new_reparto1.drop('full_name', axis=1, inplace=True)
new_reparto1

actor_id                  title
0            1       ACADEMY DINOSAUR
1            1   ANACONDA CONFESSIONS
2            1            ANGELS LIFE
3            1  BULWORTH COMMANDMENTS
4            1          CHEAPER CLYDE
...        ...                    ...
1156       196                    NaN
1157       197                    NaN
1158       198                    NaN
1159       199                    NaN
1160       200                    NaN

[1161 rows x 2 columns]

In [928]:
new_reparto2=pd.merge(peliculas, reparto, on='title', how='left')
new_reparto2

film_id              title  \
0           1   ACADEMY DINOSAUR   
1           1   ACADEMY DINOSAUR   
2           1   ACADEMY DINOSAUR   
3           1   ACADEMY DINOSAUR   
4           1   ACADEMY DINOSAUR   
...       ...                ...   
2463      997         YOUTH KICK   
2464      997         YOUTH KICK   
2465      998       ZHIVAGO CORE   
2466      999  ZOOLANDER FICTION   
2467     1000          ZORRO ARK   

                                            description  language_id  length  \
0     A Epic Drama of a Feminist And a Mad Scientist...            1      86   
1     A Epic Drama of a Feminist And a Mad Scientist...            1      86   
2     A Epic Drama of a Feminist And a Mad Scientist...            1      86   
3     A Epic Drama of a Feminist And a Mad Scientist...            1      86   
4     A Epic Drama of a Feminist And a Mad Scientist...            1      86   
...                                                 ...          ...     ...   
2463  A Touching Drama of a Teacher And a Cat who mu...            1     179   
2464  A Touching Drama of a Teacher And a Cat who mu...            1     179   
2465  A Fateful Yarn of a Composer And a Man who mus...            1     105   
2466  A Fateful Reflection of a Waitress And a Boat ...            1     101   
2467  A Intrepid Panorama of a Mad Scientist And a B...            1      50   

     rating                         special_features  category_id  \
0        PG         Deleted Scenes,Behind the Scenes            6   
1        PG         Deleted Scenes,Behind the Scenes            6   
2        PG         Deleted Scenes,Behind the Scenes            6   
3        PG         Deleted Scenes,Behind the Scenes            6   
4        PG         Deleted Scenes,Behind the Scenes            6   
...     ...                                      ...          ...   
2463  NC-17               Trailers,Behind the Scenes           12   
2464  NC-17               Trailers,Behind the Scenes           12   
2465  NC-17                           Deleted Scenes           11   
2466      R                  Trailers,Deleted Scenes            0   
2467  NC-17  Trailers,Commentaries,Behind the Scenes            0   

             full_name  
0     PENELOPE GUINESS  
1      CHRISTIAN GABLE  
2        LUCILLE TRACY  
3          SANDRA PECK  
4     PENELOPE GUINESS  
...                ...  
2463     SANDRA KILMER  
2464        VAL BOLGER  
2465          UMA WOOD  
2466               NaN  
2467               NaN  

[2468 rows x 9 columns]

In [929]:
new_reparto2.drop(['full_name', 'description', 'language_id', 'length', 'rating', 'special_features', 'category_id', 'full_name'], axis=1, inplace=True)
new_reparto2

film_id              title
0           1   ACADEMY DINOSAUR
1           1   ACADEMY DINOSAUR
2           1   ACADEMY DINOSAUR
3           1   ACADEMY DINOSAUR
4           1   ACADEMY DINOSAUR
...       ...                ...
2463      997         YOUTH KICK
2464      997         YOUTH KICK
2465      998       ZHIVAGO CORE
2466      999  ZOOLANDER FICTION
2467     1000          ZORRO ARK

[2468 rows x 2 columns]

In [930]:
new_reparto3=pd.merge(new_reparto1, new_reparto2, on='title', how='left')
new_reparto3

actor_id             title  film_id
0            1  ACADEMY DINOSAUR      1.0
1            1  ACADEMY DINOSAUR      1.0
2            1  ACADEMY DINOSAUR      1.0
3            1  ACADEMY DINOSAUR      1.0
4            1  ACADEMY DINOSAUR      1.0
...        ...               ...      ...
5602       196               NaN      NaN
5603       197               NaN      NaN
5604       198               NaN      NaN
5605       199               NaN      NaN
5606       200               NaN      NaN

[5607 rows x 3 columns]

In [931]:
new_reparto3.drop('title', axis=1, inplace=True)
new_reparto3

actor_id  film_id
0            1      1.0
1            1      1.0
2            1      1.0
3            1      1.0
4            1      1.0
...        ...      ...
5602       196      NaN
5603       197      NaN
5604       198      NaN
5605       199      NaN
5606       200      NaN

[5607 rows x 2 columns]

In [932]:
new_reparto3.duplicated().any() 

True

In [933]:
new_reparto3.drop_duplicates()

actor_id  film_id
0            1      1.0
16           1     23.0
25           1     25.0
34           1    106.0
35           1    140.0
...        ...      ...
5602       196      NaN
5603       197      NaN
5604       198      NaN
5605       199      NaN
5606       200      NaN

[1161 rows x 2 columns]

In [934]:
reparto = new_reparto3
reparto

actor_id  film_id
0            1      1.0
1            1      1.0
2            1      1.0
3            1      1.0
4            1      1.0
...        ...      ...
5602       196      NaN
5603       197      NaN
5604       198      NaN
5605       199      NaN
5606       200      NaN

[5607 rows x 2 columns]

### Guardado de csv limpios

In [935]:
actores.to_csv('../data_clean/actores_limp.csv', index=False)

In [936]:
categorias.to_csv('../data_clean/categorias_limp.csv', index=False)

In [937]:
peliculas.to_csv('../data_clean/peliculas_limp.csv', index=False)

In [938]:
inventario.to_csv('../data_clean/inventario_limp.csv', index=False)

In [939]:
idiomas.to_csv('../data_clean/idiomas_limp.csv', index=False)

In [940]:
reparto.to_csv('../data_clean/reparto_limp.csv', index=False)

In [941]:
alquileres.to_csv('../data_clean/alquileres_limp.csv', index=False)